plugins/scripting/scripting-ide-services/src/org/jetbrains/kotlin/scripting/ide_services/compiler/analyzer.kt

```diff
@@ -0,0 +1,288 @@
+import org.jetbrains.kotlin.resolve.calls.smartcasts.DataFlowInfoFactory
+import org.jetbrains.kotlin.resolve.diagnostics.Diagnostics
+import org.jetbrains.kotlin.resolve.lazy.*
+import org.jetbrains.kotlin.resolve.lazy.data.KtClassLikeInfo
+import org.jetbrains.kotlin.resolve.lazy.declarations.*
+import org.jetbrains.kotlin.resolve.scopes.ImportingScope
+import org.jetbrains.kotlin.resolve.scopes.utils.parentsWithSelf
+import org.jetbrains.kotlin.resolve.scopes.utils.replaceImportingScopes
+import org.jetbrains.kotlin.scripting.compiler.plugin.impl.KotlinResolutionFacade
+import org.jetbrains.kotlin.scripting.definitions.ScriptPriorities
+import org.jetbrains.kotlin.scripting.ide_services.util.ReplHistory
+import kotlin.script.experimental.api.SourceCode
+
+class CodeAnalyzer(private val environment: KotlinCoreEnvironment) {
+    private val container: ComponentProvider
+    private val topDownAnalysisContext: TopDownAnalysisContext
+    private val topDownAnalyzer: LazyTopDownAnalyzer
+    private val resolveSession: ResolveSession
+    private val scriptDeclarationFactory: ScriptMutableDeclarationProviderFactory
+    private val replState = ResettableAnalyzerState()
+
+    val module: ModuleDescriptorImpl
+
+    val trace: BindingTraceContext = CliBindingTrace()
+
+    init {
+        // Module source scope is empty because all binary classes are in the dependency module, and all source classes are guaranteed
+        // to be found via ResolveSession. The latter is true as long as light classes are not needed in REPL (which is currently true
+        // because no symbol declared in the REPL session can be used from Java)
+        container = TopDownAnalyzerFacadeForJVM.createContainer(
+            environment.project,
+            emptyList(),
+            trace,
+            environment.configuration,
+            environment::createPackagePartProvider,
+            { _, _ -> ScriptMutableDeclarationProviderFactory() }
+        )
+
+        this.module = container.get()
+        this.scriptDeclarationFactory = container.get()
+        this.resolveSession = container.get()
+        this.topDownAnalysisContext = TopDownAnalysisContext(
+            TopDownAnalysisMode.TopLevelDeclarations, DataFlowInfoFactory.EMPTY, resolveSession.declarationScopeProvider
+        )
+        this.topDownAnalyzer = container.get()
+    }
+
+    interface ReplLineAnalysisResult {
+        val scriptDescriptor: ClassDescriptorWithResolutionScopes?
+        val diagnostics: Diagnostics
+
+        data class Successful(
+            override val scriptDescriptor: ClassDescriptorWithResolutionScopes,
+            override val diagnostics: Diagnostics
+        ) :
+            ReplLineAnalysisResult
+
+        data class WithErrors(override val diagnostics: Diagnostics) :
+            ReplLineAnalysisResult {
+            override val scriptDescriptor: ClassDescriptorWithResolutionScopes? get() = null
+        }
+
+        data class Stateless(
```

Comment:
This one would benefit from an explanation comment

Reply:
Added a comment